File: TestLeetCode/LinkedList/Reorder.cs

// Definition for singly-linked list.
public class ListNode
{
    public int val;
    public ListNode next;
    public ListNode(int x) { val = x; }
}

public class ReorderListSolution
{
    public void ReorderList(ListNode head)
    {
        if (head == null || head.next == null) return;

        // Find the middle of the list using slow and fast pointer
        ListNode slow = head, fast = head.next;
        while (fast != null && fast.next != null)
        {
            slow = slow.next;
            fast = fast.next.next;
        }

        // Reverse the second half of the list
        ListNode second = slow.next;
        ListNode prev = null;
        slow.next = null;
        while (second != null)
        {
            ListNode tmp = second.next;
            second.next = prev;
            prev = second;
            second = tmp;
        }

        // Merge the two halves
        ListNode first = head;
        second = prev;
        while (second != null)
        {
            ListNode tmp1 = first.next;
            ListNode tmp2 = second.next;
            first.next = second;
            second.next = tmp1;
            first = tmp1;
            second = tmp2;
        }
    }
}

File: python/linked_list/reorder.py
from python.common import ListNode


class ReorderList:
    def reorder_list(self, head: ListNode | None) -> None:
        if not head or not head.next: return
        slow, fast = head, head.next
        while fast and fast.next: slow, fast = slow.next, fast.next.next
        second, slow.next = slow.next, None
        previous = None
        while second: following = second.next; second.next = previous; previous, second = second, following
        first, second = head, previous
        while second:
            first_next, second_next = first.next, second.next
            first.next, second.next = second, first_next
            first, second = first_next, second_next
